File: 7.stacks_and_queues/lesson_7_stone_wall.py
def solution(H):
    lenght = len(H)
    stack = [0]*lenght
    stack[0] = H[0]
    top_index = 1
    cubes = 1

    for index in range(1, lenght):
        if(top_index != 0):
            if(stack[top_index - 1] > H[index]):
                while(top_index != 0 and stack[top_index - 1] > H[index]):
                    top_index -= 1
                    stack[top_index] = 0
                    cubes += 1
        if(stack[top_index - 1] != H[index]):
            stack[top_index] = H[index]
            top_index += 1

    return cubes + top_index - 1

#got 100% on my own yaaay!

File: 7.stacks_and_queues/lesson_7_stone_wall.js
function solution(H) {
    let blocks = 0
    let stack = new Array()
    stack.push(H[0])

    for(let i = 1; i < H.length; i ++) {
        if(stack.length == 0) {
            stack.push(H[i])
        } else {
            let top = stack[stack.length - 1]
            while(top > H[i]) {
                stack.pop()
                blocks++
                if(stack.length == 0) {
                    break
                }
                top = stack[stack.length - 1]
            }
            if(stack.length == 0 || top < H[i]) {
                stack.push(H[i])
            }
        }
    }
    return blocks + stack.length
}